crates/profiling/src/vtune_linux.rs

```diff
@@ -0,0 +1,138 @@
+//! Profile:
+//!     amplxe-cl -run-pass-thru=--no-altstack -v -collect hotspots target/debug/wasmtime --vtune test.wasm
+//!
+//! Note: amplxe-cl is a command-line tool for Vtune which should be installed.
+
+use anyhow::Result;
+use core::ptr;
+use ittapi_rs::*;
+use std::boxed::*;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::sync::Mutex;
+use wasmtime_environ::entity::PrimaryMap;
+use wasmtime_environ::wasm::DefinedFuncIndex;
+use wasmtime_environ::Module;
+use wasmtime_runtime::VMFunctionBody;
+
+/// Interface for driving the ittapi for VTune support
+pub struct VTuneAgent {
+    // Note that we use a mutex internally to serialize state updates
+    // since multiple threads may be sharing this agent.
+    state: Mutex<State>,
+}
+
+/// Interface for driving vtune
+#[derive(Clone, Debug, Default)]
+struct State {
+    /// Unique identifier for the jitted function
+    method_id: HashMap<(usize, DefinedFuncIndex), u32>,
+}
+
+impl VTuneAgent {
+    /// Intialize a VTuneAgent and write out the header
+    pub fn new() -> Result<Self> {
+        let state = State {
+            method_id: HashMap::new(),
+        };
+        Ok(VTuneAgent {
+            state: Mutex::new(state),
+        })
+    }
+}
+
+impl State {
+    /// Return the unique method ID for use with the ittapi
+    pub fn get_method_id(&mut self, module_id: usize, func_idx: DefinedFuncIndex) -> u32 {
+        let method_id;
+        unsafe {
+            method_id = self
+                .method_id
+                .entry((module_id, func_idx))
+                .or_insert_with(|| iJIT_GetNewMethodID());
+        }
+        *method_id
+    }
+
+    /// Load module
+    pub fn event_load(
+        &mut self,
+        method_id: u32,
+        module_name: &str,
+        method_name: &str,
+        addr: *const u8,
+        len: usize,
+    ) -> () {
+        let filename = "unknown_filename";
+        let jmethod = _iJIT_Method_Load {
+            method_id: method_id,
+            method_name: CString::new(method_name)
+                .expect("CString::new failed")
+                .into_raw(),
+            method_load_address: addr as *mut ::std::os::raw::c_void,
+            method_size: len as u32,
+            line_number_size: 0,
+            line_number_table: ptr::null_mut(),
+            class_id: 0,
+            class_file_name: CString::new(module_name)
+                .expect("CString::new failed")
+                .into_raw(),
+            source_file_name: CString::new(filename)
+                .expect("CString::new failed")
+                .into_raw(),
+        };
+        let jmethod_ptr = Box::into_raw(Box::new(jmethod));
+        unsafe {
+            let _ret = iJIT_NotifyEvent(
+                iJIT_jvm_event_iJVM_EVENT_TYPE_METHOD_LOAD_FINISHED,
+                jmethod_ptr as *mut ::std::os::raw::c_void,
+            );
+        }
+    }
+
+    /// Shutdown module
+    pub fn _event_shutdown(&mut self) -> () {
```

Comment:
If this needs to be hooked up to get vtune working, I'd recommend implementing `Drop for VTuneAgent` which would call this method

Reply:
It doesn't at all ...  it seems to only have purpose if you want to stop profiling and any point while code is still executing. Having drop for VTuneAgent call this would probably be a good idea nonetheless though.  +1